k8t/templates.py

```diff
@@ -74,37 +74,65 @@ def validate(template_path: str, values: dict, engine: Environment) -> bool:
 
 
 def get_variables(template_path: str, engine: Environment) -> Set[str]:
-    template_source = engine.loader.get_source(engine, template_path)[
-        0
-    ]
+    template_source = engine.loader.get_source(engine, template_path)[0]
     abstract_syntax_tree = engine.parse(template_source)
 
+    undeclared = meta.find_undeclared_variables(abstract_syntax_tree)
+
     defaults = {
-        hasattr(filter.node, "name") and filter.node.name
+        filter.node.name
 
         for filter in abstract_syntax_tree.find_all(nodes.Filter)
 
-        if filter.name == "default"
+        if filter.name == "default" and hasattr(filter.node, "name")
+    }
+
+    guarded = {
+        test.node.name
+
+        for test in abstract_syntax_tree.find_all(nodes.Test)
+
+        if test.name == 'defined' and hasattr(test.node, "name")
+    }
+
+    local = {
+        assign.target.name
+
+        for assign in abstract_syntax_tree.find_all(nodes.Assign)
     }
 
     return (
-        meta.find_undeclared_variables(
-            abstract_syntax_tree) - (set(engine.globals.keys()) - PROHIBITED_VARIABLE_NAMES) - defaults
+        undeclared - (set(engine.globals.keys()) - PROHIBITED_VARIABLE_NAMES) - defaults - guarded - local
     )
 
 
 def get_secrets(template_path: str, engine: Environment) -> Set[str]:
-    template_source = engine.loader.get_source(engine, template_path)[
-        0
-    ]
+    template_source = engine.loader.get_source(engine, template_path)[0]
     abstract_syntax_tree = engine.parse(template_source)
 
+    def parse_call_arg(node: nodes.Node, parent: nodes.Node) -> str:
```

Comment:
It doesn't cover a lot of expression nodes, i.e. `Add`. And it probably still would be fine, but we actually don't need to evaluate `get_secret` node's value. AFAICS we find secrets only to identify if we use any secrets or not and raise error if no secrets provider configured.

So it would be simpler just to search for any occurrence of `get_secret` node during validation instead.

Reply:
that's a good point actually